# Geheime Botschaften
Zum Testen des RSA-Verfahrens haben Alexandra und Bastian folgende Schlüsselpaare erstellt:  

|                        | Alexandra             | Bastian               |
|------------------------|-----------------------|-----------------------|
| öffentlicher Schlüssel | $(e_A,n_A)=(107,187)$ | $(e_B,n_B)=(XXX,XXX)$ |
| geheimer Schlüssel     | $(d_A,n_A)=(3,187)$   | $(d_B,n_B)=(7,143)$   |


### Aufgabe 1
 Verschlüsseln Sie Alexandras Nachricht „HALLO“ an Bastian mit der Blocklänge 1 und der Codierung A=1, B=2, C=3, … 


#### Lösung

In [1]:
alphabet = " ABCDEFGHIJKLMNOPQRSTUVWXYZ"
print( alphabet.find("H"), 
       alphabet.find("A"),
       alphabet.find("L"),
       alphabet.find("L"),
       alphabet.find("O") )

8 1 12 12 15


In [2]:
print( (8**103)%143, 
       (1**103)%143,
       (12**103)%143,
       (12**103)%143, 
       (15**103)%143 )

83 1 12 12 141


Die Nachricht von Alexandra an Bastian lautet also (83, 1, 12, 12, 141)

## Aufgabe 2
Sie haben folgende Antwort von Bastian an Alexandra abgefangen: (178,135,113,53, 113,171)  
Bestimmen Sie nachvollziehbar Alexandras geheimen Schlüssel und entschlüsseln Sie damit die Botschaft (Codierung und Blocklänge wie oben).

#### Lösung
Faktorisierung von 187 durch Ausprobieren: n= 187 = 11∙17 = p∙q  

Also Eulersche $\varphi$-Funktion: 𝜑(n)=(p-1) ∙ (q-1) = 10∙16 = 160  

Berechne die modulare Inverse d modulo 160, also d mit d∙107 = 1 mod 160 mittels des erweiterten Euklidischen Algorithmus´:

$$ \begin{align}
160 &= 1 \cdot 107 + 53 \\
107 &= 2 \cdot 53 +1 \\
53  &= 53 \cdot 1 +0
\end{align} $$

$$ \begin{align}
1 &= 107  - 2 \cdot 53 \\
  &= 107 - 2 \cdot (160-1 \cdot 107) \\
  &= 3 \cdot 107 - 2 \cdot 160
\end{align} $$

Also gilt: 
$$ 1 = 1 \bmod 160 = (3 \cdot 107 - 2 \cdot 160) \bmod 160 = (3 \cdot 107) \bmod 160$$

Damit ist d = 3 gefunden und zusammen mit n= 187 lautet Alexandras geheimer Schlüssel also: 
$$(d_A,n_A)=(3,187)$$

Damit Entschlüsselung des abgefangenen Geheimcodes:

In [3]:
print( (178**3) % 187, 
       (135**3) % 187 , 
       (113**3) % 187,  
       ( 53**3) % 187, 
       (113**3) % 187, 
       (171**3) % 187 )

19 16 5 25 5 18


Also ergibt sich für den Geheimtext:

In [4]:
print( alphabet[19]+alphabet[16]+alphabet[5]+alphabet[25]+alphabet[5]+alphabet[18] )

SPEYER


### Aufgabe 3 
Warum ist eine längere verschlüsselte Nachricht mit Blocklänge 1 selbst mit den sichersten RSA-Schlüsseln relativ leicht zu knacken (→ Kryptologie 1)?

#### Lösung
Auf diese Weise werden alle Buchstaben des Alphabets immer durch jeweils die gleiche Zahl
dargestellt, es handelt sich also um eine monoalphabetische Substitution, die bei einem längeren
Text recht leicht durch eine Häufigkeitsanalyse „geknackt“ werden kann.

---
# Signatur

### Aufgabe 4
 Hier finden Sie den Beginn eines Programms, mit dem Bastian eine Nachricht an Alexandra probehalber signieren möchte. Berechnen Sie die Werte für die Ausgaben (1) bis (4).  
![Bild Quelltext](kryptologie-aufgabe-bild-quelltext.png)<br>
Anmerkung: Der Quelltext ist hier bewusst als Bild hinterlegt. Sie müssen den Quelltext nicht abtippen, es ist ausreichend, wenn sie *im Kopf* nachvollziehen, was das Programm macht. 

#### Lösung
Die berechneten Werte sind  

|||
|-|-|  
| (1) | 17 |
| (2) | 30 |
| (3) | ('HI', 30) |
| (4) | ('HO', 30) |

## Aufgabe 5
Ergänzen Sie das obige Python-Programm um die Überprüfung der Signatur durch den Empfänger inklusive einer fallabhängigen Ausgabe, ob die Nachricht vermutlich unverändert übertragen wurde oder nicht.

### Lösung
Der zu ergänzende codeblock lautet:
```
# Überprüfen der empfangenen Nachricht
hashEmpfangeneNachricht = hashBerechnen(text)
hashEntschluesselt = modularePotenz(hashVerschluesselt, oeffentlich)
if hashEmpfangeneNachricht == hashEntschluesselt:
    print("Die Nachricht wurde vermutlich nicht verändert.")
else:
    print("Die Nachricht wurde verändert.")
```

Hier kommt das komplette Python-Programm zum Testen:

In [5]:
def zahl(c):
    """
    >>> zahl('A')
    1
    """
    if c == ' ':
        return 0
    else:
        return ord(c)-65+1

''' Die folgende Funktion berechnet die modulare Potenz,
    je nach Einsatz und übergebenem Schlüssel dient sie dabei dem Verschlüsseln,
    dem Entschlüsseln, dem Signieren oder dem Überprüfen der Signatur.'''
def modularePotenz(zahl, schluessel):
    return (zahl**schluessel[0])%schluessel[1]

def hashBerechnen(text):
    summe = 0
    for zeichen in text:
        # zahl(zeichen) bestimmt die Stelle im Alphabet, also zahl("A")=1, zahl("B")=2 usw.
        summe = summe + zahl(zeichen) 
    return summe % 27

e = 103
d = 7
n = 143
oeffentlich = (e,n) # Bastians öffentlicher Schlüssel
geheim = (d,n) # Bastians geheimer Schlüssel

# Erstellen der signierten Nachricht
text = 'HI'
hashText = hashBerechnen(text)
print("(1)", hashText) # (1)
hashVerschluesselt = modularePotenz(hashText, geheim)
print("(2)", hashVerschluesselt) # (2)
nachricht = (text, hashVerschluesselt)
print("(3)", nachricht) #(3)

# Veränderung des Nachrichtentextes "unterwegs"
text = 'HO'
nachrichtEmpfangen = (text, hashVerschluesselt)
print("(4)", nachrichtEmpfangen) # (4)

# Überprüfen der empfangenen Nachricht
hashEmpfangeneNachricht = hashBerechnen(text)
hashEntschluesselt = modularePotenz(hashVerschluesselt, oeffentlich)
if hashEmpfangeneNachricht == hashEntschluesselt:
    print("Die Nachricht wurde vermutlich nicht verändert.")
else:
    print("Die Nachricht wurde verändert.")

(1) 17
(2) 30
(3) ('HI', 30)
(4) ('HO', 30)
Die Nachricht wurde verändert.


### Aufgabe 6
Handelt es sich bei der verwendeten Hash-Funktion um eine gute Wahl? Begründen Sie. 

#### Lösung

Die Hashfunktion ist nicht geschickt gewählt, da alle möglichen Nachrichtentexte nur auf 27 Hash-Werte abgebildet werden. Die Chance, dass ein veränderter Nachrichtentext zufälligerweise oder bei einem Angriff auch absichtlich auf die gleiche Zahl führt, ist also sehr hoch (-> schwache Kollisionsresistenz).